#4257 Added functionality to create previews for post links using openn graph data from those links.

debanshuk opened the pull request:
#### Summary
Removed the existing implementation of website link (URL) preview generation which was using Noembed service (a oEmbed provider) and implemented preview generation using Open Graph meta data present on webpages.

Note: Preview generation for youtube videos still uses existing implementation which make them playable in the preview itself. That hasn't been changed. 

#### Ticket Link
Github issue: https://github.com/mattermost/platform/issues/4257

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Added API documentation (required for all new APIs) (Link to PR for API reference update: https://github.com/mattermost/mattermost-api-reference/pull/64)
- [x] All new/modified APIs include changes to the drivers
- [x] Has UI changes
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/platform/blob/master/i18n/en.json) and [.../webapp/i18n/en.json](https://github.com/mattermost/platform/tree/master/webapp/i18n/en.json)) updates

debanshuk commented:
Build is failing for this pull request with following error:
api/post.go:23:2: cannot find package "github.com/dyatlov/go-opengraph/opengraph" in any of:
	/home/ubuntu/workspace/mattermost-platform-pr/src/github.com/mattermost/platform/vendor/github.com/dyatlov/go-opengraph/opengraph (vendor tree)
	/usr/local/go/src/github.com/dyatlov/go-opengraph/opengraph (from $GOROOT)
	/home/ubuntu/workspace/mattermost-platform-pr/src/github.com/dyatlov/go-opengraph/opengraph (from $GOPATH)

Though 'github.com/dyatlov/go-opengraph/opengraph' is present int vendor directory. I have also tried to rebuild after adding 'oepngraph' as a subpackage of 'github.com/dyatlov/go-opengraph' but the build is stil falling.

All tests and style checks are passing on my local machine. Could anyone help in figuring out the reason of build failure?

debanshuk commented:
Build is failing for this pull request with following error:
api/post.go:23:2: cannot find package "github.com/dyatlov/go-opengraph/opengraph" in any of:
	/home/ubuntu/workspace/mattermost-platform-pr/src/github.com/mattermost/platform/vendor/github.com/dyatlov/go-opengraph/opengraph (vendor tree)
	/usr/local/go/src/github.com/dyatlov/go-opengraph/opengraph (from $GOROOT)
	/home/ubuntu/workspace/mattermost-platform-pr/src/github.com/dyatlov/go-opengraph/opengraph (from $GOPATH)

Though 'github.com/dyatlov/go-opengraph/opengraph' is present int vendor directory. I have also tried to rebuild after adding 'oepngraph' as a subpackage of 'github.com/dyatlov/go-opengraph' but the build is stil falling.

All tests and style checks are passing on my local machine. Could anyone help in figuring out the reason of build failure?

debanshuk commented:
Create pull request on mattermost-api-reference with API documentation for /get_opengraph_metadata API: https://github.com/mattermost/mattermost-api-reference/pull/64

debanshuk commented:
I deleted and go-opengraph package from vendor directory and installed in again and an now the build is passing. I guess I had added the package by manually copying and pasting it from go src folder where I had installed it originally for doing some testing with it.

Build is working fine when I installed the package using glide install. I am new in go, so didn't knew this can cause problem.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0a474373208766838.spinmint.com:8065/pr4890

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0a474373208766838

it33 commented:
Thanks @debanshuk, really appreciate your help here. 

Just testing this now and hit an issue that's preventing me from testing further, could you take a look? 

#### 1) Account Settings > Advanced > Preview pre-release features disappears after activating link preview feature

Here's what it should look like (2 features): 

![image](https://cloud.githubusercontent.com/assets/177788/21476065/b9c2f560-cae8-11e6-828b-e340ef222b4f.png)

With the PR, after checking "Show experimental previews..." the setting disappears: 

![image](https://cloud.githubusercontent.com/assets/177788/21476024/2547f372-cae8-11e6-8f45-dea90e04e081.png)

Link previews is a tricky feature and we need the ability for users to turn it on and off depending on their preference.

debanshuk commented:
@it33, I have intentionally removed the experimental link preview setting from advance settings, as I thought that that is the desired behaviour after reading the ticket. But it looks like I missed removing the option when it's value is false from the display. But as asked in your comment, I can reinstate the option.

Though as stated in the second point of your comment the link preview setting in display section only works for gif previews (actually for all images and YouTube videos). So, should I tie the website link preveiws too with this setting (which I guess I should have done earlier) and make sure experimental link preview setting in advance setting is removed completely and properly? I guess that would be better option than having two settings.

it33 commented:
Thanks @debanshuk 

1. Please include your link preview with the Display > Link preview setting. 

This makes the feature consistent with other previews. 

2. Please include your link preview as an option in the Advanced Settings page, as the feature will be considered experimental in its first release. 

This is because a) the feature will have a pretty huge impact and b) it's difficult to comprehensively test. If it is just one of these factors, it's easier to release directly, but because it is both, we need to keep it under Advanced for a release or two. 

I've assigned this PR to @esethna to test later this week and have proposed some additional test cases: https://github.com/mattermost/platform/pull/4899/files

Highly appreciate your work, I think people will love this :)

debanshuk commented:
@it33, Sure, I'll reinstate the option in advance settings and also tie it with Display > Link preview setting.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-046ac0d9ad3178126.spinmint.com:8065/pr4890

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-046ac0d9ad3178126

debanshuk commented:
@it33 @esethna, I have reinstated experimental link preview option in advance settings and have tied expansion of link preview with Display > Link preview setting.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server destroyed

esethna commented:
Thanks @debanshuk! When you get a chance can you please help to resolve the conflicts so we can fire up a test server?

mattermod commented:
Spinmint test server created at: http://i-07d05739f4bfc522e.spinmint.com:8065/pr4890

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-07d05739f4bfc522e

debanshuk commented:
@esethna, I have rebased the branch resolving conflicts.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0b337df8b7a73230b.spinmint.com:8065/pr4890

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0b337df8b7a73230b

jasonblais commented:
Sidhant is helping do a first pass on this PR, followed by a PM

jasonblais commented:
Sidhant tested, moving testing to @jasonblais next

jasonblais commented:
@debanshuk I've been testing this out and have to say that you've done a really great work here! Wow!

A few notes from first phase of testing:

1 - the following sites don't provide a link preview, although expected:
    - Wikipedia sites, e.g. https://en.wikipedia.org/wiki/Olympus_Mons
    - http://www.techmeme.com/
    - https://about.gitlab.com

2 - after generating a link preview containing an image, the center panel doesn't scroll all the way to the bottom, but stays about half-way through the image. Example shown below, reproduced for each link preview

![image](https://cloud.githubusercontent.com/assets/13119842/21746203/2bbd220e-d50b-11e6-8c2d-92112f610a16.png)

Expected: Scrolls to the bottom

![image](https://cloud.githubusercontent.com/assets/13119842/21746251/4096811a-d50c-11e6-8dfb-30ad9cb60b75.png)


3 - The Globe and Mail news articles ([sample link](http://www.theglobeandmail.com/news/national/three-canadians-shortlisted-for-global-teacher-prize/article33429901/)) provide an image preview of their logo instead of the one in the news article 

![image](https://cloud.githubusercontent.com/assets/13119842/21746226/d15b13a6-d50b-11e6-9feb-faec0bba9a41.png)

I'm not sure if this is just how that particular website is set up with respect to Open Graph, but I did try in other apps where the correct image was displayed.

![image](https://cloud.githubusercontent.com/assets/13119842/21746230/e27c25bc-d50b-11e6-89e8-41736cbce669.png)

4 - Links from Discourse give a broken image preview ([sample link](https://forum.mattermost.org/t/building-without-docker/2677))

![image](https://cloud.githubusercontent.com/assets/13119842/21746244/09f1c0b6-d50c-11e6-9e41-3bbb30534df4.png)

Expected: This image is shown in other apps with link previews. Not sure again if it's a bug with Open Graph, fine to fix in a separate PR. 

![image](https://cloud.githubusercontent.com/assets/13119842/21746254/5b71de76-d50c-11e6-9efb-0dfa54498c7f.png)

I'll queue the following items for our UX meeting next Monday to discuss (do not worry about these yet, I'll post a follow-up by Tuesday):
5 - Should link previews be displayed on the right-hand side (comment thread)?
6 - Should we display a blue bar for link previews? I feel it's unnecessary since we surround link previews with a box, and it also clashes with the reply bar.
7 - Link previews with formatted text (bold, italic, strikethrough).

Other notes (do not worry about these in this pull request):
8 - The /collapse and /expand slash commands work for all image links in all channels, both for new and old ones - expected is that they work for past messages in the channel. Existing ticket: https://mattermost.atlassian.net/browse/PLT-4380
9 - Status indicator shown for replies in mobile: https://mattermost.atlassian.net/browse/PLT-5194
10 - Single collapsed link previews don't stay collapsed after refresh. Existing ticket: https://mattermost.atlassian.net/browse/PLT-3638

jasonblais commented:
@debanshuk -- tickets created for items 5 --> 10, none which blocks the PR:
5 - https://mattermost.atlassian.net/browse/PLT-4380
6 - https://mattermost.atlassian.net/browse/PLT-5194
7 - https://mattermost.atlassian.net/browse/PLT-3638
8 - https://mattermost.atlassian.net/browse/PLT-5268
9 - https://mattermost.atlassian.net/browse/PLT-5269
10 - https://mattermost.atlassian.net/browse/PLT-5270

Not sure if you've had a chance to take a look at 1 --> 4?

debanshuk commented:
@jasonblais, Sorry for late reply, didn't got  change to look at 1 --> 4 earlier. Following are my replies to each issues:

1 - https://en.wikipedia.org/wiki/Olympus_Mons and http://www.techmeme.com/ don't have any OpenGraph meta data on their pages, and as stated in the issue description is a site doesn't have OG description we shouldn't show generate it's preview.
"If a link posted has an open graph description in the web page metadata, show a preview"
"If there is no description on the web page, do not create a link preview."

https://about.gitlab.com/ (Gitlab) does have OpenGraph metadata on their page and also have og:description but for some reason it's not getting picked by the OG lib which we are using. I'll look into the issue.

2 - I'll look into the issue. I should stay at bottom showing the whole image if it's already at bottom, but looks like it's not working that way for some reason.

3 - Yes **Account Settings > Display > Link** Preview only affects the section which can be expanded/collapsed. In the description it says "Expand links to show a preview of content, when available.". Also, even if someone turn this setting off they can expand their link's preview by clicking the expand icon (
![screenshot from 2017-01-13 15-09-37-mm](https://cloud.githubusercontent.com/assets/2102907/21925180/562051f6-d9a2-11e6-9849-49f8a16b73dd.png)
) or via `/expand` command. In youtube video or image link previews the expand/collapse icon is at the top of the message hence the whole preview is expanded/collapsed

Collapsed:
![screenshot from 2017-01-13 15-06-03-mm](https://cloud.githubusercontent.com/assets/2102907/21925094/e137ad80-d9a1-11e6-81e7-036569adb3d6.png)

Expanded:
![screenshot from 2017-01-13 15-05-17-mm](https://cloud.githubusercontent.com/assets/2102907/21925084/d884fc9c-d9a1-11e6-9dad-4a318ee227c3.png)
 
If you want the whole website link preview to be expanded/collapsed when the setting is turn on/off then we would have to move the expand/collapse (
![screenshot from 2017-01-13 15-08-43-mm](https://cloud.githubusercontent.com/assets/2102907/21925163/334abb58-d9a2-11e6-84b2-90e5553b8ca4.png)
) icon to the top of the message.

The requirement of keeping the icon only at image is specified in the issue itself.
"If the website preview contains an image, make only the image portion of the website preview collapsable.

Show the collapse icon beside the description."

Currently, if someone want to completely turn off website link preview they can do it from **Account Settings > Advance > Preview pre-release features > Show experimental previews of link content, when available **.

4.1 -  This is because Globe and Mail are clumsy, they have this image URL http://static.theglobeandmail.ca/a04/news/national/article33429900.ece/ALTERNATES/w620/teachers-prize-Doucet1225nw1.JPG their OG meta data but have this URL https://sec.images.theglobeandmail.com/media/www/images/flag/gam-square.png as Secure-URL (with HTTPS) in OG meta data. And we are giving more preference to secure URL than plain HTTP URL. 

Should I change this behaviour give preference ti plain HTTP URL? This would solve the issue on Globe and Mail, but won't be an ideal solution for the problem.

4.2 This is happening because https://github.com/mattermost/platform/pull/4890 has stated "/images/default-apple-touch-icon.png" in it's og:image meta tag. "/images/default-apple-touch-icon.png" is URL without domain and scheme and per OpenGraph's documentation og:image or any URL type should have a URL with http:// or https:// protocols (Ref: http://ogp.me/#url. Definition of og:image says "og:image - An image URL which should represent your object within the graph."). So, the correct value should be "https://forum.mattermost.org/images/default-apple-touch-icon.png".

So, I'll look in into 1 and 2; need a reply from you on 3 and 4.1 and don't need to do anything on 4.2, as it's an issue in OpenGraph implementation at https://forum.mattermost.org/.

jasonblais commented:
Thanks @debanshuk, much appreciated!

1 - Sounds good for https://en.wikipedia.org/wiki/Olympus_Mons and http://www.techmeme.com/ pages - I thought they had OG description. 

Thanks for looking into the GitLab one.

2 - +1, thanks for looking into it

3 - The **Account Settings > Display > Link Preview** setting help text reads `Expand links to show a preview of content, when available.` Hence, it's related to links, such as showing the text preview for website links.

The /expand and /collapse shortcuts apply for image links only, not for text previews. So using these shortcuts should only control whether an image preview is collapsed or expanded within the link preview, and shouldn't be tied with the Account Setting option.
 - note: [we have an open improvement ticket to improve how the /expand and /collapse shortcuts work](https://mattermost.atlassian.net/browse/PLT-4380).

In summary, as an end user, I would expect the following:
  1. definitions:
     a. Non-collapsible content: Links that do not have expand/collapse arrows. Includes text for website link previews.
     b. Collapsible content: All links that can be expanded/collapsed with the arrows. Includes image links within website previews, plain images, GIFs (such as http://i.giphy.com/l3vR3gvEdsdJl26oU.gif) and videos (such as YouTube videos)
 2. **Account Settings > Display > Link Preview** is on: 
     a. non-collapsible content: shows a preview. 
           - Website link previews for OG data only show if the Advanced setting is selected (since this is the first release, as [explained by it33 above](https://github.com/mattermost/platform/pull/4890#issuecomment-269164190)). 
           - Images within the website previews can be collapsed/expanded using the arrows or /expand and /collapse shortcuts
     b. collapsible content: show a preview by default. Can be collapsed/expanded using the arrows or /expand and /collapse shortcuts
 3. **Account Settings > Display > Link Preview** is off: 
     a. non-collapsible content: no preview shown and there is no way to expand the content (both text and images).  
     b. collapsible content: don't show the preview by default but the user can expand it using the expand arrow or the /expand command.

What do you think? Any thoughts or feedback? 

Right now, 2a, 2b and 3b are implemented but 3a is missing. Not sure if this would be feasible to add for this PR.

4.1 - We can leave this for now. If we're pulling the correct OG meta data, then that's good. Also agree that giving preference to plain HTTP URLs is not the ideal solution.

4.2 - 

>  This is happening because #4890

^did you mean to reference another link/page? I agree it's not related to this PR, Mattermost team should go and fix the URL to https://forum.mattermost.org/images/default-apple-touch-icon.png

debanshuk commented:
Thanks @jasonblais for the replies. Following are my replies on point number 3 and 4.2.

**3 -** Well, behaviours 3.a and 3.b are bit different from each other (as in 3.b users can expand the preview content even if they have Link Preview setting turned off, but in case of 3.a they can't) and I thing it can be confusing for users it same setting applies differently on different content. 

Though If you want I have implement the want stated in 3.a in this PR itself. Should I do that?

**4.2 -** Yes I meant to reference another link/page. I have update my comment with the correct link. It's the same sample link which you had provided in your comment, link to a Mattermost forum page https://forum.mattermost.org/t/building-without-docker/2677 .

jasonblais commented:
Thanks @debanshuk for the reply! 

Let me queue item number 3 for Monday's UX meeting to discuss with others as well? You do bring up a good point about inconsistency, will have to think about that carefully.

In the meanwhile maybe you can help look at 1 (for GitLab page) and 2?

debanshuk commented:
@jasonblais, Sure. 

I have figured out the reason of issue 1 (GitLab issue), it because they have incorrect meta tag for og:description their page. Following this meta tag present on their page:

```
<meta content="Code, test &amp;amp; deploy with GitLab. Everyone can contribute!" name="og:description">
```
 It has "og:description" text in `name` attribute. It should have been in `property` attributed as stated here: http://ogp.me/. So, it's an issue at GitLabs end.

I am still looking into point number 2. Would update here when it's resolved.

Also, I found another issue after I rebased my branch from lasted master branch. This issue isn't due to my changes, as I can reproduce it on master branch too. Following is the issue:

"""
Any setting present in **Account Settings > Advance > Preview pre-release features** isn't persisted after saving them and reloading the page. As default value of each setting is false (ie, disabled), enabling any of these settings only has effect for the currently and the effect is lost after app is reloaded.

Also, the new setting Ie. "Enable the ability to make and receive one-on-one WebRTC calls" disappears after it is enabled and saved.
"""

debanshuk commented:
@jasonblais, I have added a fix for point number 2, so that the post list stays at the bottom (if it's at the bottom before the preview is added) and whole preview is visible. Please re-review the functionality.

jasonblais commented:
Awesome, thanks @debanshuk! I'll get back to you on 3 later today

jasonblais commented:
Hey @debanshuk, 

This was discussed and I understand the case better now -- the **Account Settings > Display > Link Preview** setting was designed to work for image links only, which is why they are controlled by the `/expand` and `/collapse` commands. It is not really intended to work for website link previews.

So what you currently have is correct -- the website link previews you're adding should only be controlled by the setting in **Account Settings > Advance > Preview pre-release features**.

However, propose we make a small update to the UI text on the settings modal so that 3) is more clear to users - the UI text is currently under review, so I'll let you know as soon as it's approved.

Otherwise we should be all good! :) 

---- 

Also, big thanks for finding the bugs with the pre-release features. Tickets files here:
 - https://mattermost.atlassian.net/browse/PLT-5302
 - https://mattermost.atlassian.net/browse/PLT-5230

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0dc9fdf0c286a30c9.spinmint.com:8065/pr4890

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0dc9fdf0c286a30c9

jasonblais commented:
We're still reviewing the text, but that shouldn't block this PR.

I re-tested on Windows 10, Chrome:
   - carried out tests described in https://github.com/mattermost/platform/tree/master/tests (they will be updated based on the findings in this PR, e.g. wikipedia pages don't have any OpenGraph meta data so no preview is expected)
      - includes testing on center pane, right-hand side, search
   - also tested on direct message channels
   - also tested the bug where channel wouldn't stick to the bottom after link preview is generated in different scenarios

Apart from the following bug and improvement tickets already files and not caused by this PR, didn't find anything new. Excellent work @debanshuk! 

https://mattermost.atlassian.net/browse/PLT-4380
https://mattermost.atlassian.net/browse/PLT-5194
https://mattermost.atlassian.net/browse/PLT-3638
https://mattermost.atlassian.net/browse/PLT-5268
https://mattermost.atlassian.net/browse/PLT-5269
https://mattermost.atlassian.net/browse/PLT-5270
https://mattermost.atlassian.net/browse/PLT-5302
https://mattermost.atlassian.net/browse/PLT-5230

mattermod commented:
Spinmint test server destroyed

jwilander commented on webapp/components/post_view/components/post_attachment_opengraph.jsx:
```diff
@@ -0,0 +1,192 @@
+        const bestImageLte = nearestPointData.nearestPointLte;  // Best image <= 150px height and width
+
+        const finalBestImage = (
+            !$.isEmptyObject(bestImageLte) && bestImageLte.height <= this.imageDimentions.height && bestImageLte.width <= this.imageDimentions.width ?
```
We're trying to move away from using jQuery for a few reasons. Would you be able to add a utils function that checks for an empty object without using jQuery and use that here instead?

I think EMCA5 method `Object.keys(obj).length === 0 && obj.constructor === Object` should work

debanshuk replied:
@jwilander, Sure I'll do that.

jwilander commented on webapp/components/post_view/components/post_attachment_opengraph.jsx:
```diff
@@ -0,0 +1,192 @@
+                                            /> : ''
+                                        }
+                                    </div>
+                                    {imageUrl && this.state.imageVisible ?
```
This is minor, but can you not use the tertiary operator for conditionally including JSX and just a regular if statement? I think it's fine to use for the props, but as soon as tertiary operators go multi-line they become much more difficult to read

debanshuk commented:
@jwilander, I have done the changes stated by you in review and have also rebased the branch from latest master branch.

jwilander commented:
Looks good, thanks!

enahum commented on webapp/actions/global_actions.jsx:
```diff
@@ -596,3 +596,18 @@ export function redirectUserToDefaultTeam() {
     }
 }
+
+export function updateOpenGraphMetaData(url) {
```
Why is this function called `updateOpenGraphMetaData` when it's using `Client.getOpenGraphMetaData` also the word `MetaData` shouldn't be `Metadata` => `getOpenGraphMetadata`?

debanshuk replied:
@enahum, I'll change it to `requestOpenGraphMetadata` and also change `MetaData` to `metadata` everywhere. The word 'update' actually meant 'request for update in store', but that is confusing.

enahum commented on api/post.go:
```diff
@@ -649,3 +652,29 @@ func getFileInfosForPost(c *Context, w http.ResponseWriter, r *http.Request) {
 	}
 }
+
+func getOpenGraphMetaData(c *Context, w http.ResponseWriter, r *http.Request) {
```
should be named `getOpenGraphMetadata`?

enahum commented on api/post_test.go:
```diff
@@ -1298,3 +1299,33 @@ func TestGetPermalinkTmp(t *testing.T) {
 	}
 }
+
+func TestGetOpenGraphMetaData(t *testing.T) {
```
same here

enahum commented on webapp/client/client.jsx:
```diff
@@ -27,6 +27,8 @@ export default class Client {
             unknownError: 'We received an unexpected status code from the server.'
         };
+
+        this.openGraphMetaDataOnGoingRequests = {};  // Format: {<url>: true}
```
and here :P

enahum commented on webapp/client/client.jsx:
```diff
@@ -1767,6 +1769,30 @@ export default class Client {
+        // There can be only one on going request for one url.
+        // This helps where there are multiple posts with same URL are rendered simultaneously.
+        var self = this;
+        if (!this.openGraphMetaDataOnGoingRequests[url]) {
```
In the event there is more than one simultaneous post with the same URL what happens? I don't see success being called for that case, is the post rendering metadata?

debanshuk replied:
@enahum, Currently the this function is only called from https://github.com/mattermost/platform/pull/4890/files/68587e43d1b4e38a9f47ee69994b36fcb217b875#diff-f1f7963b47ead3e8ded08b371f14420fR600. And in case of simultaneous requests for same URL, success function of only one request would be called. Which is fine as the success function triggers and action to update OG data, which update update it for all posts and they would be rendered with data received from the URL.

Though I understand that as this is a client function it can be called with success/error functions, in which case only the first request's success/error function would be called. I can move logic of rejecting more than one request with same URL to 'updateOpenGraphMetaData' action instead.

enahum commented on webapp/stores/opengraph_store.jsx:
```diff
@@ -0,0 +1,66 @@
+}
+
+var OpenGraphStore = new OpenGraphStoreClass();
+OpenGraphStore.setMaxListeners(40);  // Not expecting more that 40 post previews rendered at a time
```
we do fetch 60 posts at a time, you may want to change this to 60?

debanshuk commented:
@enahum, I have changed all usage of MetaData to Metadata (have also update constants where only DATA was used instead of METADATA) and have moved handling of `openGraphMetaDataOnGoingRequests` from client to action. I have also renamed `updateOpenGraphMetaData` action to `requestOpenGraphMetadata`. Please re-review the updated PR.

Let me know if there is anything else.

jasonblais commented:
Thanks @debanshuk for the contribution! This is a great one :)
